### PR TITLE
Add genetic map auto detection and conversion

### PR DESCRIPTION
**Pull request by LouisLeNezet**
## PR checklist

Closes #216 
Closes #150 

Genetic map has been added to `test_batch`, `test`, `test_quilt`, `test_stitch`, `test_panelprep`, `test_all` and `test_vcf`. Therefore the md5sum for the test `- with map` have been updated.
The test names have been clarified.

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/phaseimpute/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/phaseimpute _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core pipelines lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: docs/usage.md [hunk 1]
```diff
@@ -129,6 +129,57 @@ chr21	16609476	A,G
...
+
+The following parameters are automatically detected, but can also be set (e.g. when you only provide `pos` and `cm` with no header):
+
+- `--map_sep`: Field separator used in the map file (e.g. "\t", " ", ",")
+- `--map_header`: Whether the file contains a header row (`true` or `false`)
+- `--map_colnames`: Ordered list of column names in the file
+
+For the example below, the map file uses tab separators, contains a header, and provides the columns in the following order: `chr`, `id`, `cm`, `pos`. Therefore, the appropriate parameters would be:
+
+`--map_sep "\t" --map_header true --map_colnames "chr,id,cm,pos"`
```
Comment (atrigila) [inline docs/usage.md]: This is extremely complicated for an user and the fact that it has more parameters to set configurations implies more risk of errors and the necessity of more tests and error messages. 

I would suggest that we stick to one map format  (e.g. glimpse format), then require users to upload that format as input file and then if they need to run any tool, the autoconversion will be from that strict format to the rest of the tools. The one map format to use could be the one that is the most "comprehensive" of them all (that it includes all the necessary information). We could then point out to external documentation on how to obtain those. 

The autoconversion can be a single module that produces all the types of formats as different outputs, then used by the tools.

File: modules/local/mapautodetect/main.nf
Comment (atrigila) [inline modules/local/mapautodetect/main.nf:1]: Before implementing mapautodetect and mapconvert locally, check if we can use the GAWK nf-core module. If this nf-core module does not fit your purpose, consider developing a general‐purpose module under nf-core CUSTOM/* modules and also use template scripts if it has more than 20 lines. You can then import the module into this pipeline via nf‑core modules install, which will avoid having a local copy.

File: docs/usage.md [hunk 2]
```diff
@@ -129,6 +129,57 @@ chr21	16609476	A,G
...
+
+```csv title="chr21.map"
+chr1	id1	0.00000	55550
+chr1	id2	0.00000	632942
+chr1	id3	0.00000	633147
+chr1	id4	0.41029	785910
+chr1	id5	0.41742	788439
+chr1	id6	0.41764	788511
+chr1	id7	0.43061	792862
+chr1	id8	0.43586	794568
```
Comment (atrigila) [inline docs/usage.md]: You should validate the map structure with a schema, which columns are mandatory? which are optional? Correct errors should be raised.

File: nextflow_schema.json [hunk 3]
```diff
@@ -264,6 +264,27 @@
...
+                    "fa_icon": "fas fa-heading",
+                    "hidden": true
+                },
+                "map_colnames": {
+                    "type": "string",
+                    "description": "Column names for the genetic map file.",
+                    "default": null,
+                    "fa_icon": "fas fa-columns",
+                    "pattern": "^(chr|id|cm|pos)(,(chr|id|cm|pos))*$",
+                    "hidden": true
```
Comment (atrigila) [inline nextflow_schema.json]: This can be detected with a schema.

File: workflows/phaseimpute/tests/test_stitch.nf.test
Comment (atrigila) [inline workflows/phaseimpute/tests/test_stitch.nf.test:1]: Do not change tests. Add new tests if necessary, but do not change previous tests. Otherwise, we cannot evaluate if previous functionality remains stable.

File: docs/usage.md (hunk 1 above)
Comment (atrigila) [reply docs/usage.md]: I think the format should be Oxford 3-column format

File: subworkflows/local/map_detect_convert_gawk/main.nf
Comment (atrigila) [inline subworkflows/local/map_detect_convert_gawk/main.nf:1]: If no nf-core modules fulfill this, this could be replaced by a single custom module doing the detect + convert using python/R (and submitted to nf-core). The AWK code has no validation, no numeric type enforcement, no header awareness, it is a bit harder to unit test and fragile to whitespaces.

File: conf/test_glimpse2.config
Comment (atrigila) [inline conf/test_glimpse2.config:1]: Please do not change the original tests, this messes up with the nf-tests later and it is more difficult to do traceability. We can manage tests cases using the params in nf-test.

**Review by atrigila — CHANGES_REQUESTED** [review]

File: conf/test_glimpse2.config
Comment (LouisLeNezet) [reply conf/test_glimpse2.config:1]: Here, I only reordered the input to match the other config files.

The aim of the test from my point of view is to be the closer as possible to real scenario.
That's why I did add a map file in test_batch, test, test_quilt, test_stitch, test_panelprep, test_all and test_vcf. As they weren't using one yet.

File: workflows/phaseimpute/tests/test_stitch.nf.test
Comment (LouisLeNezet) [reply workflows/phaseimpute/tests/test_stitch.nf.test:1]: I made sure that the md5sum only changed for the test concerned by the addition of the map file.
The only snapshot that changed are for test where we added a map.

This is also why I've changed the test for a more descriptive name to better follow the changes.

File: docs/usage.md (hunk 1 above)
Comment (LouisLeNezet) [reply docs/usage.md]: This could indeed be a simpler solution, where we would only need `MAPCONVERT`.
But as a consequence we push the burden of the conversion back to the user if they don't have the right format.

For which format to use as default, I think it could be nice to ask the nf-core community. When I worked on dogs, the genetic map file I found was mostly in 4-column plink format.

File: modules/local/mapautodetect/main.nf
Comment (LouisLeNezet) [reply modules/local/mapautodetect/main.nf:1]: Using the `GAWK` modules would make the maintenance complicated if we pass the program as a string.
But we could pass it as a file from the assets directory.

File: docs/usage.md (hunk 2 above)
Comment (LouisLeNezet) [reply docs/usage.md]: The schema usage, could indeed be a good solution.
I will try to implement it.

File: conf/test_glimpse2.config
Comment (LouisLeNezet) [reply conf/test_glimpse2.config:1]: I can remove it from test_batch and test_vcf as they are not aimed to test the tools.
But for me, the test aimed at tool testing should have the map to me more comprehensive.

File: subworkflows/local/map_detect_convert_gawk/main.nf
Comment (LouisLeNezet) [reply subworkflows/local/map_detect_convert_gawk/main.nf:1]: Done

**Comment by LouisLeNezet** [comment]
Hi @atrigila,
I've updated the pipeline to use a custom nf-core modules that does the detection as well as the conversion in one go.
The tests are now using the genetic map file to homogenize all the test and make it clearer which use what.
Is it okay for you ?

**Comment by atrigila** [comment]
I will give a new review and let you know, thanks!

File: conf/test_beagle5.config [hunk 4]
```diff
@@ -27,12 +27,15 @@ params {
...
-    map          = "${projectDir}/tests/csv/map_plink.csv"
-
     // Genome references
     fasta = params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz"
 
+    // Map file
+    map          = "${projectDir}/tests/csv/map_plink.csv"
+    map_sep      = " "
+    map_header   = false
+    map_colnames = "chr,id,cm,pos"
```
Comment (atrigila) [inline conf/test_beagle5.config]: You should remove these parameters. The module should be able itself to detect if the file has a separator, a header and the column names, given that these are standard for all the different types of genetic files.

File: conf/test_glimpse2.config [hunk 5]
```diff
@@ -28,13 +28,13 @@ params {
     // Genome references
     fasta  = params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz"
     panel  = "${projectDir}/tests/csv/panel.csv"
+    map    = "${projectDir}/tests/csv/map_glimpse.csv"
```
Comment (atrigila) [inline conf/test_glimpse2.config]: I understand your point of view of reorganizing the tests so that they look alike or that using a map is more comprehensive. You can do these changes later, in a separate PR. However, right now we need to ensure that:
1. The original tests do not change their results.
2. You can add a second test in nf-test adding the map. 

The current state does not make it simple to reviewers to realize if the md5sum changes are just because the map was added or because of another reason, besides the original test disappearing. So please, keep all the original tests as-is and add the new ones as variations in nf-test.

File: conf/test_minimac4.config [hunk 6]
```diff
@@ -23,24 +23,20 @@ params {
...
 
     // Input data
-    input  = "${projectDir}/tests/csv/sample_vcf_snp.csv"
+    input        = "${projectDir}/tests/csv/sample_vcf_snp.csv"
+    input_region = "${projectDir}/tests/csv/region.csv"
 
     // Genome references
     fasta  = params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz"
     panel  = "${projectDir}/tests/csv/panel.csv"
+    map    = "${projectDir}/tests/csv/map_glimpse.csv"
```
Comment (atrigila) [inline conf/test_minimac4.config]: Same comments as for `test_glimpse2`

File: conf/test_panelprep.config [hunk 7]
```diff
@@ -26,6 +26,7 @@ params {
     fasta          = params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz"
     input_region   = "${projectDir}/tests/csv/region.csv"
     panel          = "${projectDir}/tests/csv/panel.csv"
+    map            = "${projectDir}/tests/csv/map_glimpse.csv"
```
Comment (atrigila) [inline conf/test_panelprep.config]: Same comments as for `test_glimpse2`

File: conf/test_quilt.config [hunk 8]
```diff
@@ -28,6 +28,7 @@ params {
 
     // Genome references
     fasta   = params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz"
+    map     = "${projectDir}/tests/csv/map_glimpse.csv"
```
Comment (atrigila) [inline conf/test_quilt.config]: Same comments as for `test_glimpse2`

File: conf/test_stitch.config [hunk 9]
```diff
@@ -28,6 +28,9 @@ params {
 
     // Genome references
     fasta   = params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz"
+    map     = "${projectDir}/tests/csv/map_glimpse.csv"
```
Comment (atrigila) [inline conf/test_stitch.config]: Same comments as for `test_glimpse2`

File: docs/usage.md [hunk 10]
```diff
@@ -129,6 +129,36 @@ chr21	16609476	A,G
...
+A typical genetic map samplesheet contains one row per chromosome.
+For example, a file specifying maps for two chromosomes might look like this:
+
+```console title="map.csv"
+chr,map
+chr21,GRCh38_chr21.glimpse.map
+chr22,GRCh38_chr22.glimpse.map
+```
+
+Genetic map files are available in various formats, so an autoconversion module has been implemented. A valid map file for `nf-core/phaseimpute` must:
```
Comment (atrigila) [inline docs/usage.md]: Which formats are available? Can you list them?
The autoconversion module accepts all formats? If not, which formats does it accept as input and which does it convert into? 
Clarify that the autoconversion is bundled in the pipeline.

File: docs/usage.md [hunk 11]
```diff
@@ -129,6 +129,36 @@ chr21	16609476	A,G
...
+
+- be a text file (compressed or uncompressed),
+- contain at least the following two columns:
+  - `pos`: physical position (base pairs)
+  - `cm`: genetic position (centiMorgans)
+- additional optional columns include:
+  - `chr`: chromosome name
+  - `id`: variant identifier
```
Comment (atrigila) [inline docs/usage.md]: You can also add a table here with all the formats available and the "structure" they have. 

Additionally, add a link to a sample (such as to test-datasets).

File: docs/usage.md (hunk 10 above)
Comment (atrigila) [reply docs/usage.md]: Clarify which are the outputs that are produced by the autoconversion module.

File: tests/default.nf.test.snap [hunk 12]
```diff
@@ -118,6 +123,17 @@
...
+                    "initialisation/map_convertion",
+                    "initialisation/map_convertion/1000GP_chr21.glimpse.map",
+                    "initialisation/map_convertion/1000GP_chr21.minimac.map",
+                    "initialisation/map_convertion/1000GP_chr21.plink.map",
+                    "initialisation/map_convertion/1000GP_chr21.stitch.map",
+                    "initialisation/map_convertion/1000GP_chr22.glimpse.map",
+                    "initialisation/map_convertion/1000GP_chr22.minimac.map",
+                    "initialisation/map_convertion/1000GP_chr22.plink.map",
+                    "initialisation/map_convertion/1000GP_chr22.stitch.map",
+                    "initialisation/map_convertion/versions.yml",
```
Comment (atrigila) [inline tests/default.nf.test.snap]: Should we have a `versions.yml` here since we are using topic channels?

File: workflows/phaseimpute/tests/test_all.nf.test [hunk 13]
```diff
@@ -8,7 +8,36 @@ nextflow_pipeline {
 
     config "./nextflow.config"
 
-    test("Check test_all") {
+    test("Check test_all - no map") {
```
Comment (atrigila) [inline workflows/phaseimpute/tests/test_all.nf.test]: This is exactly what I meant in my comment in `test_glimpse2` but in this case it would be the "with map". The original test config should remain the same (at least for this PR, then you can switch them).

File: conf/test_glimpse2.config (hunk 5 above)
Comment (atrigila) [reply conf/test_glimpse2.config]: This makes the reviewing process more difficult because there are more chanes and more files to review.

File: workflows/phaseimpute/main.nf [hunk 14]
```diff
@@ -119,6 +122,19 @@ workflow PHASEIMPUTE {
     def region_count = ch_region.map{ _meta, region -> region }
         .count()
 
+    ch_map_branched = ch_map
+        .branch { meta, map_file ->
+            non_empty: map_file
+            empty: true
+        }
```
Comment (atrigila) [inline workflows/phaseimpute/main.nf:129]: Do we know what happens if the `--map` param contains incomplete records (such as input rows only for chr21 but not for chr22)? Does it raise an error or warning?

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R:1]: This is a huge improvement over the original implementation in awk, thank you! There are some aspects that still need some working.

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+# Function to detect separator
+detect_separator <- function(first_line) {
+  counts <- c(
+    tab = str_count(first_line, "\\t"),
+    comma = str_count(first_line, ","),
+    semicolon = str_count(first_line, ";"),
+    space = str_count(first_line, " ")
+  )
```
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R]: Instead of implementing custom functions, `fread()` already supports automatic separator detection, automatic header detection, direct URL input, and compressed files, such as:
```
map_df <- data.table::fread(
  file_path,
  sep = "auto",
  header = "auto",
  showProgress = FALSE
)
```

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+
+  separators <- c("\\t", ",", ";", " ")
+  separators[which.max(counts)]
+}
+
+# Function to detect header
+detect_header <- function(first_line) {
+  header_keywords <- c(
+    "pos", "cm", "snp",
```
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R]: `detect_header()` uses "snp" as evidence that the first row is a header, but detect_column_names() never maps "snp" to "id".

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+  if (endsWith(file_path, ".gz")) {
+    con <- gzcon(con)
+  }
+
+  on.exit(close(con), add = TRUE)
+  readLines(con, n = 1, warn = FALSE)
+}
+
+# Function to detect separator
+detect_separator <- function(first_line) {
```
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R]: Switch to fread option as this does not take into account if delimiters appear inside quoted fields, or if you have whitespace-delimited files with repeated spaces or of the first line is a comment or if the header contains many spaces but the real delimiter is tab/comma/semicolon.

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+    header_keywords, function(keyword) {
+      grepl(keyword, first_line, ignore.case = TRUE)
+    }
+  ))) {
+    TRUE
+  } else {
+    FALSE
+  }
+}
+
```
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R:133]: The normalization logic is useful, but incomplete. Common variants like chrom, chromosome, bp, marker, rsid are not handled. That's why it is also useful to document in the `usage.md` all of the types of genetic maps that are accepted as input and with their examples.

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+    )
+    if (any(map_df[["diff"]] > tolerance)) {
+      print(map_df[map_df[["diff"]] > tolerance, ])
+      stop("cm[n] must equal cm[n-1] + ( (pos[n] - pos[n-1]) * rate[n-1])")
+    }
+  }
+
+  if (!is.numeric(map_df[["rate"]])) {
+    stop("Error: pos column should be numeric")
+  }
```
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R:230]: This is an error, right? This is checking rate, not pos?

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+    stop("Error: Position or cM missing")
+  }
+
+  if (!is.null(chr) && any(map_df[["chr"]] != chr)) {
+    stop("Error: mismatch between chr given and the chr present in file")
+  }
+
+  # Order position to ensure all successive rows have increasing position
+  map_df <- map_df[order(map_df[["pos"]]), ]
+  stopifnot(all(diff(map_df[["pos"]]) > 0))
```
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R]: I think this might give a difficult user-facing error if there are duplicated positions

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+library(data.table)
+library(stringr)
+
+# Function to parse external arguments
+parse_args <- function(x) {
+  args_list <- unlist(strsplit(x, " ?--")[[1]])[-1]
+  args_vals <- lapply(
+    args_list,
+    function(x) scan(text = x, what = "character", quiet = TRUE)
+  )
```
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R:42]: This can lead to some errors with--flag=value style args, repeated args, missing values or malformed numeric tolerance values. There are some examples on how to handle this in modules such as deseq2, which I think are used as example when developing template files for nf-core modules.

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+  writeLines(
+    paste(
+      map_df[["chr"]], map_df[["id"]],
+      map_df[["cm"]], map_df[["pos"]],
+      sep = " "
+    ),
+    con_plink
+  )
+  writeLines(
```
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R]: The final output ewriting is verbose and less safe than fwrite()

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R (hunk 18 above)
Comment (atrigila) [reply modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R:133]: Perphas something like:


normalize_names <- function(x) {
  x <- tolower(x)
  x <- gsub("[^a-z0-9#]+", "_", x)

  dplyr::case_when(
    x %in% c("chr", "#chr", "chrom", "chromosome") ~ "chr",
    x %in% c("id", "snp", "marker", "rsid") ~ "id",
    x %in% c("pos", "position", "bp") ~ "pos",
    x %in% c("cm", "genetic_map", "genetic_map_cm") ~ "cm",
    x %in% c("rate", "combined_rate", "cm_cb", "cm_mb") ~ "rate",
    TRUE ~ x
  )
}

File: modules/nf-core/custom/geneticmapconvert/tests/main.nf.test
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/tests/main.nf.test:1]: Right now the tests are mostly end-to-end snapshots which are difficult to evaluate as we cannot see the output. I would add a second layer with tiny synthetic fixtures and explicit content assertions for the generated files, especially for: cM normalization to zero, rate computation, ID preservation, chromosome injection, output column ordering. That gives you much better bug observation.

Comment (atrigila) [reply modules/nf-core/custom/geneticmapconvert/tests/main.nf.test:1]: Some ideas to improve coverage of tests:
- Add a test for a 3-column no-header STITCH-like file to see if we have the failure.
- Add a test that includes: duplicate positions, unsorted positions, single-row input, zero/negative intervals
- Add test where the file contains multiple chromosomes and meta.chr is absent. I think right now that likely passes, and it probably should not. Cases when Missing chr column, file has no chr, meta.chr provided? and file has no chr, meta.chr missing?
- Edge cases: blank/comment line before header
- Add tests for malformed tolerance values, e.g. --tolerance abc.
- Separators: You are not testing comma or semicolon-delimited inputs at all, even though the script claims to auto-detect them.

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+  writeLines("#chr\tposition\tGenetic_Map(cM)", minimac_file)
+  writeLines("position COMBINED_rate.cM.Mb. Genetic_Map.cM.", stitch_file)
+
+  # Write data to files
+  con_glimpse <- file(glimpse_file, open = "a")
+  con_minimac <- file(minimac_file, open = "a")
+  con_plink <- file(plink_file, open = "a")
+  con_stitch <- file(stitch_file, open = "a")
+
+  # Write data to files
```
Comment (atrigila) [inline modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R]: The 4 output files are built with repeated writeLines() + file connections. That works, but it is verbose and less safe than `fwrite()`.

File: modules/nf-core/custom/geneticmapconvert/tests/main.nf.test
Comment (atrigila) [reply modules/nf-core/custom/geneticmapconvert/tests/main.nf.test:1]: Right now as.numeric(tolerance) can become NA, which can later lead to confusing behavior.

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R
Comment (atrigila) [reply modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R:1]: The script has too much manual parsing and that is where most of the fragility comes from

**Review by atrigila — CHANGES_REQUESTED** [review]
Good job, just some details to polish.

File: conf/test_glimpse2.config (hunk 5 above)
Comment (LouisLeNezet) [reply conf/test_glimpse2.config]: I will do so !

File: tests/default.nf.test.snap (hunk 12 above)
Comment (LouisLeNezet) [reply tests/default.nf.test.snap]: This is one of the limitation of template() usage.
For the moment we have to emit this file.

File: workflows/phaseimpute/main.nf (hunk 14 above)
Comment (LouisLeNezet) [reply workflows/phaseimpute/main.nf:129]: This is taken care of by the `PIPELINE_INITIALISATION` sbwf.
It only output all the channel for the joint of all channels by chromosomes.
https://github.com/nf-core/phaseimpute/blob/b0adfc2331527c1630b7a1be4c6e8c9599f73dfe/subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf#L376-L424

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R
Comment (LouisLeNezet) [reply modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R:1]: As it is a nf-core modules, I will open a PR to resolve your comments and ask your review on it.

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R (hunk 21 above)
Comment (LouisLeNezet) [reply modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R:42]: I've copy paste the exact code of deseq2 actually.
What kind of malformed value would you want to guard against ?

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R (hunk 15 above)
Comment (LouisLeNezet) [reply modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R]: If that works this is nice indeed !

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R (hunk 20 above)
Comment (LouisLeNezet) [reply modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R]: I will add a warning in case of duplicate.

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R (hunk 19 above)
Comment (LouisLeNezet) [reply modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R:230]: Bad copy/paste

File: modules/nf-core/custom/geneticmapconvert/tests/main.nf.test
Comment (LouisLeNezet) [reply modules/nf-core/custom/geneticmapconvert/tests/main.nf.test:1]: Done for:
-  3-column no-header STITCH-like
- duplicate positions, unsorted positions
- multiple chromosomes and meta.chr is absent
- file has no chr, meta.chr provided
- file has no chr, meta.chr missing
- blank/comment line before header
- --tolerance abc (options(warn = 2))
- other separator

File: modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R
Comment (LouisLeNezet) [reply modules/nf-core/custom/geneticmapconvert/templates/geneticmapconvert.R:1]: I should have answer all your comments on PR https://github.com/nf-core/modules/pull/11050

File: conf/test_beagle5.config (hunk 4 above)
Comment (LouisLeNezet) [reply conf/test_beagle5.config]: Done

File: conf/test_minimac4.config (hunk 6 above)
Comment (LouisLeNezet) [reply conf/test_minimac4.config]: Done

File: conf/test_panelprep.config (hunk 7 above)
Comment (LouisLeNezet) [reply conf/test_panelprep.config]: Done

File: conf/test_quilt.config (hunk 8 above)
Comment (LouisLeNezet) [reply conf/test_quilt.config]: Done

File: conf/test_stitch.config (hunk 9 above)
Comment (LouisLeNezet) [reply conf/test_stitch.config]: Done

File: docs/usage.md (hunk 10 above)
Comment (LouisLeNezet) [reply docs/usage.md]: Done

File: docs/usage.md (hunk 11 above)
Comment (LouisLeNezet) [reply docs/usage.md]: Done

File: workflows/phaseimpute/tests/test_all.nf.test (hunk 13 above)
Comment (LouisLeNezet) [reply workflows/phaseimpute/tests/test_all.nf.test]: Resolved

File: docs/usage.md [hunk 24]
```diff
@@ -129,6 +129,46 @@ chr21	16609476	A,G
...
+
+- be a text file (compressed or uncompressed)
+- separated by comma, semicolon, space or tabulation
+- one file by chromosome
+- contain at least the following two columns (with `_` being a non alphanumeric character):
+  - `pos`: physical position (base pairs), can be named: pos, position or bp
+  - `cm`: genetic position (centiMorgans), can be named: id, snp, marker or rsid
+- additional optional columns include:
+  - `chr`: chromosome name, can be named: \_chr, chrom or chromosome
+  - `id`: variant identifier, can be named: cm or genetic_map_cm
```
Comment (atrigila) [inline docs/usage.md]: Is this an issue?:

- `id`: variant identifier, can be named: **cm** 
- - `cm`: genetic position (centiMorgans), **can be named: id**

File: docs/usage.md [hunk 25]
```diff
@@ -129,6 +129,46 @@ chr21	16609476	A,G
...
+- if no header provided:
+  - if three columns should be chr, pos, cm (minimac format)
```
Comment (atrigila) [inline docs/usage.md]:
```suggestion
  - if three columns should be, strictly in this order: **chr, pos, cm** (minimac format)
```

File: docs/usage.md [hunk 26]
```diff
@@ -129,6 +129,46 @@ chr21	16609476	A,G
...
+  - if four columns should be chr, id, cm, pos (plink format)
```
Comment (atrigila) [inline docs/usage.md]:
```suggestion
  - if four columns should be, strictly in this order: **chr, id, cm, pos** (plink format)
```

File: docs/usage.md
Comment (atrigila) [inline docs/usage.md:1]: You might want to update `output.md` as well with the output files produced?

File: workflows/phaseimpute/main.nf [hunk 27]
```diff
@@ -119,6 +122,19 @@ workflow PHASEIMPUTE {
...
+
+    CUSTOM_GENETICMAPCONVERT(ch_map_branched.non_empty)
```
Comment (atrigila) [inline workflows/phaseimpute/main.nf:131]: My only question is if this should be in pipeline_initialisation / genome preparation subworkflow.

File: docs/usage.md (hunk 24 above)
Comment (LouisLeNezet) [reply docs/usage.md]: Sorry I copy paste in the wrong place.

Comment (LouisLeNezet) [reply docs/usage.md]: Done

File: workflows/phaseimpute/main.nf (hunk 27 above)
Comment (LouisLeNezet) [reply workflows/phaseimpute/main.nf:131]: This is something I plan to do in a separate PR in a subworkflow named `PREPARE_GENOME`.

File: docs/usage.md
Comment (LouisLeNezet) [reply docs/usage.md:1]: Done

**Review by atrigila — APPROVED** [review]
